### PR TITLE
[JAX] Refine MHA API and add DPA API

### PR DESCRIPTION
**Pull request by zlsh80826**
This PR does the following changes:

1. Refine the MHA API to align with TransformerLayer API
    a. Replace `num_heads` with `num_attention_heads`.
    b. Replace `droout_rate` with `attention_dropout`.
    c. Replace `output_layernorm` with `input_layernorm`. where this params is used to apply the layernorm on the inputs.
    d. Replace `apply_residual_connection_post_layernorm` with `return_layernorm_output`.
    e. Replace `fuse_qkv` with `fused_qkv_params`.
    f. The old params are marked as deprecated, and will be removed in the future.
2. Add DotProductAttention module API
    Since JAX doesn't have the [view/stride](https://pytorch.org/docs/stable/generated/torch.Tensor.view.html) concept, all tensors are always in continuous memory format. To have a unify API for all `qkvpacked`, `kvpacked`, `qkv separate` with 1 input tensor, 2 input tensors and 3 input tensors respectively. This module accepts `qkv_layout` to parse `query`, `key`, `value` tensors for the different inputs combinations.
    
        * bs3hd: query tensor is treated as a qkvpacked tensor with shape = [b, s, 3, h, d].
          key and value arguments in :attr:`__call__()` are ignored in this layout.
        * bshd_bs2hd: query tensor with shape = [b, s, h, d]. key tensor is treaded as a kvpacked
          tensor with shape = [b, s, 2, h, d]. `value` argument in :attr:`__call__()` is ignored.
        * bshd_bshd_bshd: query, key, and value are seperated with shape = [b, s, h, d].
3. Support `BSHD_BSHD_BSHD` fused attention custom call

### CODE REVIEW

File: transformer_engine/jax/cpp_extensions.py [hunk 1]
```diff
@@ -2662,6 +2664,445 @@ def cross_fused_attn_bwd(q: jnp.ndarray, kv: jnp.ndarray, bias: jnp.ndarray,
...
+
+        *q_batch_shape, q_max_seqlen, num_heads, q_head_dim = q_aval.shape
+        *kv_batch_shape, kv_max_seqlen, num_gqa_groups, kv_head_dim = k_aval.shape
+        assert q_batch_shape == kv_batch_shape
+        assert q_head_dim == kv_head_dim
+        assert k_aval.shape == v_aval.shape
+        out_aval = q_aval.update(shape=q_aval.shape, dtype=q_dtype)
+
+        # backend determines the softmax buffer shape/dtype
+        backend = FusedAttnHelper(q_dtype, k_dtype, NVTE_QKV_Layout.NVTE_BSHD_BS2HD, attn_bias_type,
```
Comment (denera) [inline transformer_engine/jax/cpp_extensions.py:2700]: Should the QKV layout for this call be `NVTE_QKV_Layout.NVTE_BSHD_BSHD_BSHD`?

File: transformer_engine/jax/csrc/modules.cpp [hunk 2]
```diff
@@ -1488,5 +1487,265 @@ void CrossFusedAttnBackward(cudaStream_t stream, void **buffers, const char *opa
     nvte_tensor_pack_destroy(&aux_input_tensors);
 }
 
+pybind11::tuple GetFusedAttnForwardWorkspaceSizes(
```
Comment (denera) [inline transformer_engine/jax/csrc/modules.cpp:1490]: The only difference I can see between this function and `GetCrossFusedAttnForwardWorkspaceSizes()` is the choice of `qkv_layout`. Is that correct?

If so, we should combine all the workspace functions into a single one, which would also make it much easier to eliminate duplicate code in the JAX Python primitives.

File: transformer_engine/jax/csrc/modules.cpp [hunk 3]
```diff
@@ -1488,5 +1487,265 @@ void CrossFusedAttnBackward(cudaStream_t stream, void **buffers, const char *opa
...
+                        dbias_tensor.data(), q_cu_seqlens_tensor.data(),
+                        kv_cu_seqlens_tensor.data(), q_max_seqlen, kv_max_seqlen, scaling_factor,
+                        dropout_probability, qkv_layout, bias_type, mask_type,
+                        query_workspace_tensor.data(), nullptr);
+
+    auto work_shape = MakeShapeVector(query_workspace_tensor.shape());
+    return pybind11::make_tuple(work_shape, query_workspace_tensor.dtype());
+}
+
+void FusedAttnBackward(cudaStream_t stream, void **buffers, const char *opaque, size_t opaque_len) {
```
Comment (denera) [inline transformer_engine/jax/csrc/modules.cpp:1661]: Same as other comments, I think the only difference between the different fused attention kernels is what comes out of the buffers. We can combine them by storing the QKV format in the descriptor and using it to determine how it reads the buffers (packed QKV, packed KV, or unpacked).

File: transformer_engine/jax/cpp_extensions.py [hunk 4]
```diff
@@ -2662,6 +2664,445 @@ def cross_fused_attn_bwd(q: jnp.ndarray, kv: jnp.ndarray, bias: jnp.ndarray,
                                                            is_training=is_training)
 
 
+class FusedAttnFwdPrimitive(BasePrimitive):
```
Comment (denera) [inline transformer_engine/jax/cpp_extensions.py:2667]: I'm seeing a lot of boilerplate replication between `SelfFusedAttn`, `CrossFusedAttn` and `FusedAttn` primitives.

I think we should consider implementing a base primitive class like `BaseFusedAttn` and then specialize into the `QKVPackedFusedAttn`, `KVPackedFusedAttn`, `UnpackedFusedAttn` subtypes (similar to how `Softmax` primitive specializes into `Scaled`, `ScaledMasked` and `ScaledUpperTriangMasked`) to make this more maintainable.

File: transformer_engine/jax/praxis/transformer.py [hunk 5]
```diff
@@ -59,30 +62,114 @@ def __call__(self, q_seqlen: JTensor, k_seqlen: JTensor, bidirectional: bool = T
...
     scale_attn_logits: bool = False
     scaled_query_init: bool = True
     float32_logits: bool = False
 
+    # Deprecated parameters
+    num_heads: int | None = None
+    dropout_rate: float | None = None
+    output_layernorm: bool | None = None
+    apply_residual_connection_post_layernorm: bool | None = None
+    fuse_qkv: bool | None = None
```
Comment (ksivaman) [inline transformer_engine/jax/praxis/transformer.py:146]:
```suggestion
    num_heads: Optional[int] = None
    dropout_rate: Optional[float] = None
    output_layernorm: Optional[bool] = None
    apply_residual_connection_post_layernorm: Optional[bool] = None
    fuse_qkv: Optional[bool] = None
```

File: transformer_engine/jax/praxis/transformer.py [hunk 6]
```diff
@@ -59,30 +62,114 @@ def __call__(self, q_seqlen: JTensor, k_seqlen: JTensor, bidirectional: bool = T
         return self.relative_position_bias(q_seqlen, k_seqlen, bidirectional)
 
 
+class DotProductAttention(TransformerEngineBaseLayer):
+    """DotProductAttention"""
+
+    head_dim: int = 64
+    num_attention_heads: int = 16
+    num_gqa_groups: int | None = None
+    attention_dropout: float = 0.
```
Comment (ksivaman) [inline transformer_engine/jax/praxis/transformer.py:71]: Discussion: This is the default in TE's DPA API in PyTorch and JAX MHA API, but I'm not sure if this is the best default? Would also need fixing in the other APIs.

File: qa/L0_jax_unittest/test.sh [hunk 7]
```diff
@@ -4,6 +4,8 @@
 
 set -xe
 
+export NVTE_FUSED_ATTN=1
+
 : ${TE_PATH:=/opt/transformerengine}
 pytest -Wignore -v $TE_PATH/tests/jax -k 'not distributed'
```
Comment (ksivaman) [inline qa/L0_jax_unittest/test.sh:10]:
```suggestion
: ${TE_PATH:=/opt/transformerengine}
NVTE_FUSED_ATTN=1 pytest -Wignore -v $TE_PATH/tests/jax -k 'not distributed'
```

File: transformer_engine/common/fused_attn/fused_attn.cpp [hunk 8]
```diff
@@ -105,8 +105,8 @@ NVTE_Fused_Attn_Backend nvte_get_fused_attn_backend(
     bool flag_m512 = false;
     bool flag_arb = false;
     if ((sm_arch_ == 80 || sm_arch_ == 90)
-            && (max_seqlen_q <= 512)
-            && (max_seqlen_kv <= 512)
+            && (max_seqlen_q <= 512 && max_seqlen_q % 64 == 0)
```
Comment (denera) [inline transformer_engine/common/fused_attn/fused_attn.cpp:108]: @zlsh80826 You might have already tested this locally but we should trigger the PyTorch and Paddle CIs just to have a record of clean results saved on GitHub before merging.

File: transformer_engine/common/fused_attn/fused_attn.cpp [hunk 9]
```diff
@@ -105,8 +105,8 @@ NVTE_Fused_Attn_Backend nvte_get_fused_attn_backend(
...
+            && (max_seqlen_kv <= 512 && max_seqlen_kv % 64 == 0)
```
Comment (denera) [inline transformer_engine/common/fused_attn/fused_attn.cpp:109]: @zlsh80826 You might have already tested this locally but we should trigger the PyTorch and Paddle CIs just to have a record of clean results saved on GitHub before merging.

File: transformer_engine/jax/csrc/modules.cpp (hunk 2 above)
Comment (denera) [inline transformer_engine/jax/csrc/modules.cpp:1490]: The only difference I can see between this function and `GetCrossFusedAttnForwardWorkspaceSizes()` is the choice of `qkv_layout`. Is that correct?

If so, we should combine all the workspace functions into a single one, which would also make it much easier to eliminate duplicate code in the JAX Python primitives.

File: transformer_engine/jax/csrc/modules.cpp [hunk 10]
```diff
@@ -1488,5 +1487,265 @@ void CrossFusedAttnBackward(cudaStream_t stream, void **buffers, const char *opa
...
+                        q_cu_seqlens_tensor.data(), kv_cu_seqlens_tensor.data(),
+                        dummy_rng_state_tensor.data(), q_max_seqlen, kv_max_seqlen, is_training,
+                        scaling_factor, dropout_probability, qkv_layout, bias_type, mask_type,
+                        query_workspace_tensor.data(), nullptr);
+
+    auto work_shape = MakeShapeVector(query_workspace_tensor.shape());
+    return pybind11::make_tuple(work_shape, query_workspace_tensor.dtype());
+}
+
+void FusedAttnForward(cudaStream_t stream, void **buffers, const char *opaque, size_t opaque_len) {
```
Comment (denera) [inline transformer_engine/jax/csrc/modules.cpp:1533]: Same as other comments, I think the only difference between the different fused attention kernels is what comes out of the buffers. We can combine them by storing the QKV format in the descriptor and using it to determine how it reads the buffers (packed QKV, packed KV, or unpacked).

File: transformer_engine/jax/cpp_extensions.py (hunk 4 above)
Comment (denera) [inline transformer_engine/jax/cpp_extensions.py:2667]: I see lot of boilerplate replication between `SelfFusedAttn`, `CrossFusedAttn` and `FusedAttn` primitives.

We should consider implementing a base primitive class like `BaseFusedAttn` and then specialize into the `QKVPackedFusedAttn`, `KVPackedFusedAttn`, `UnpackedFusedAttn` subtypes (similar to how `Softmax` primitive specializes into `Scaled`, `ScaledMasked` and `ScaledUpperTriangMasked`) to make this more maintainable.

File: transformer_engine/jax/cpp_extensions.py (hunk 1 above)
Comment (denera) [inline transformer_engine/jax/cpp_extensions.py:2700]: Should the QKV layout for this call be `NVTE_QKV_Layout.NVTE_BSHD_BSHD_BSHD`?

File: transformer_engine/jax/praxis/transformer.py (hunk 5 above)
Comment (zlsh80826) [reply transformer_engine/jax/praxis/transformer.py:146]: [Optional[int] is equalvant to int | None](https://docs.python.org/3/library/typing.html#typing.Optional), is it because for an unified coding style?

File: transformer_engine/jax/praxis/transformer.py (hunk 6 above)
Comment (zlsh80826) [reply transformer_engine/jax/praxis/transformer.py:71]: I will remove the default of `head_dim` and `num_attetnion_heads` here and also `MultiHeadAttention`'s [default](https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/jax/flax/transformer.py#L353-L354) for those two parameters in `TE-JAX-FLAX`. I checked that there is no default for those two parameters in [TE-PyT](https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/pytorch/attention.py#L2315-L2316) and [TE-JAX-FLAX](https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/jax/flax/transformer.py#L353-L354) DPA/MHA API.

File: transformer_engine/jax/cpp_extensions.py (hunk 1 above)
Comment (zlsh80826) [reply transformer_engine/jax/cpp_extensions.py:2700]: Yes, thanks for catching it.

**Review by denera — APPROVED** [review]

File: qa/L0_jax_unittest/test.sh (hunk 7 above)
Comment (zlsh80826) [reply qa/L0_jax_unittest/test.sh]: Fix in fcf6c564b0d1fd13b834f7f5861df89d199c1976. I moved the env to the specific test files.

File: transformer_engine/jax/praxis/transformer.py (hunk 5 above)
Comment (ksivaman) [reply transformer_engine/jax/praxis/transformer.py:149]: Yes

File: transformer_engine/jax/praxis/transformer.py (hunk 6 above)
Comment (zlsh80826) [reply transformer_engine/jax/praxis/transformer.py:71]: Fix in 8666d48a3da5f7220514646d8eb0d8c48c068bc9. Praxis layers are not allowed the no-default members, so I give the default value = 0 and check if it is set in the `setup()`, which is the same way used in the praxis repo ([set zero](https://github.com/google/praxis/blob/main/praxis/layers/attentions.py#L1149) and [check](https://github.com/google/praxis/blob/main/praxis/layers/attentions.py#L1234)).

File: tests/jax/test_praxis_layers.py [hunk 11]
```diff
@@ -40,6 +43,19 @@
 FP8_FORMATS = [Format.E4M3, Format.HYBRID]
 
 
+@pytest.fixture(autouse=True, scope='module')
+def enable_fused_attn():
+    """
+    Enable fused attn for hopper+ arch.
+    Fused attn kernels on pre-hopper arch are not deterministic.
```
Comment (cyanguwa) [inline tests/jax/test_praxis_layers.py:50]: I think this might be wrong here. For pre-hopper archs, kernels are deterministic, and for hopper+, the workspace_optimization path gives deterministic answers. This is for arbitrary seqlen backend. The max512 backend is always deterministic.

File: tests/jax/test_praxis_layers.py [hunk 12]
```diff
@@ -671,6 +689,82 @@ def test_forward(self, data_shape, dtype, attrs, rtol=1e-05, atol=1e-08):
...
+        SCALE_FACTOR: 0.125,
+    }, {
+        ATTN_MASK_TYPE: 'padding_causal',
+        TRANSPOSE_BS: False,
+        SCALE_FACTOR: 2.,
+    }, {
+        ATTN_MASK_TYPE: 'causal',
+        TRANSPOSE_BS: False,
+        SCALE_FACTOR: 1.,
+    }]
```
Comment (cyanguwa) [inline tests/jax/test_praxis_layers.py:721]: Do we want to add 'no_mask' here?

File: transformer_engine/jax/cpp_extensions.py [hunk 13]
```diff
@@ -2662,6 +2664,445 @@ def cross_fused_attn_bwd(q: jnp.ndarray, kv: jnp.ndarray, bias: jnp.ndarray,
...
+register_primitive(FusedAttnFwdPrimitive)
+
+
+def fused_attn_fwd(q: jnp.ndarray, k: jnp.ndarray, v: jnp.ndarray, bias: jnp.ndarray,
+                   q_seqlen: jnp.ndarray, kv_seqlen: jnp.ndarray, seed: jnp.ndarray,
+                   attn_bias_type: NVTE_Bias_Type, attn_mask_type: NVTE_Mask_Type,
+                   scaling_factor: float, dropout_probability: float, is_training: bool):
+    """
+    Wrapper for TE fused attention fwd, where query, key, value are seperated tensors
+    Return BMM1 -> (PreBias) -> ScaleMaskSoftmax -> (PostBias) -> (Dropout) -> BMM2
```
Comment (cyanguwa) [inline transformer_engine/jax/cpp_extensions.py:2858]: PostBias is probably after Scale and before MaskSoftmax.

**Review by cyanguwa — APPROVED** [review]

File: tests/jax/test_praxis_layers.py (hunk 12 above)
Comment (zlsh80826) [reply tests/jax/test_praxis_layers.py:725]: Add in 2b0bf343b205456bcdfd309f30fe5f4f628dec63

File: transformer_engine/jax/cpp_extensions.py (hunk 13 above)
Comment (zlsh80826) [reply transformer_engine/jax/cpp_extensions.py]: Fixed in 1d26d02a. Thanks for catching it.

File: transformer_engine/jax/praxis/transformer.py (hunk 5 above)
Comment (zlsh80826) [reply transformer_engine/jax/praxis/transformer.py]: ok, fixed in 6f660d5a.

**Review by cyanguwa — APPROVED** [review]
LGTM

**Comment by denera** [comment]
Merging this PR. Thanks @zlsh80826!
